journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,283 @@
+
+  case class AllocatedObject[F[_], O](value: O, deallocate: F[Unit], state: AllocatedObject.State)
+
+  object AllocatedObject {
+    sealed abstract class State(val isFree: Boolean)
+
+    object State {
+      case object Used extends State(false)
+      case class Free(releasedAt: FiniteDuration) extends State(true)
+    }
+  }
+
+  case class State[F[_], O](
+    objects: Map[ObjectId, AllocatedObject[F, O]],
+    waiters: Queue[Waiter[F, O]],
+  ) {
+
+    def findFreeObject: Option[(ObjectId, AllocatedObject[F, O])] =
+      objects.collectFirst {
+        case (id, obj) if obj.state.isFree => (id, obj)
+      }
+
+    def markObjectAsUsed(id: ObjectId, obj: AllocatedObject[F, O]): State[F, O] =
+      State(
+        objects + (id -> obj.copy(state = AllocatedObject.State.Used)),
+        waiters
+      )
+
+    def markObjectAsFree(id: ObjectId, obj: AllocatedObject[F, O], releasedAt: FiniteDuration): State[F, O] =
+      State(
+        objects + (id -> obj.copy(state = AllocatedObject.State.Free(releasedAt))),
+        waiters
+      )
+
+    def addObject(id: ObjectId, obj: AllocatedObject[F, O]): State[F, O] =
+      State(
+        objects + (id -> obj),
+        waiters
+      )
+
+    def removeObject(id: ObjectId): State[F, O] =
+      State(
+        objects - id,
+        waiters
+      )
+
+    def addWaiter(waiter: Waiter[F, O]): State[F, O] =
+      State(
+        objects,
+        waiters.enqueue(waiter)
+      )
+
+    def takeWaiter: (State[F, O], Option[Waiter[F, O]]) =
+      waiters.headOption match {
+        case Some(waiter) =>
+          State(objects, waiters.tail) -> waiter.some
+        case None =>
+          State(objects, Queue.empty) -> none
+      }
+  }
+
+  def fixedSize[F[_]: Async, O](
+    poolSize: Int,
+    acquireTimeout: FiniteDuration,
+    idleTimeout: FiniteDuration,
+    log: Log[F],
+  )(
+    resource: Resource[F, O]
+  ): Resource[F, ResourcePool[F, O]] = {
+
+    def poolImpl(mutex: Mutex[F], stateRef: Ref[F, State[F, O]], objCounter: Ref[F, Int]): ResourcePool[F, O] = {
+
+      val allocateObject: F[(ObjectId, AllocatedObject[F, O])] =
+        for {
+          id <- objCounter.updateAndGet(_ + 1)
+          a <- resource.allocated
+          (obj, release) = a
+        } yield id -> AllocatedObject(obj, release, AllocatedObject.State.Used)
+
+      def handleNewWaiter(waiter: Waiter[F, O]): F[Unit] = {
+        val timeout = waiter.complete(
+          Left(new TimeoutException(s"Resource acquisition timed out"))
+        )
+
+        stateRef.get.flatMap { state =>
+          state.findFreeObject match {
+            case Some((id, obj)) =>
+              for {
+                _ <- waiter.complete((id, obj.value).asRight)
+                _ <- stateRef.update(_.markObjectAsUsed(id, obj))
+              } yield ()
+
+            case None =>
+              if (state.objects.size < poolSize) {
+                for {
+                  a <- allocateObject
```

Comment:
one more reason to get rid of mutex, we block the whole thing while allocating/deallocating resource, and this would definitely be a problem